syntax = "proto3";

package worker;

service WorkerService {
  // RPC for sending a message to a channel !! Most likely can be removed!
  rpc SendChannelMessage(LobbyMessage) returns (Status) {}

  // RPC for getting channel messages (server-side streaming) !! Most likely can be removed!
  rpc GetChannelMessages (ChannelMessageRequest) returns (stream Message) {}

  // RPC to join a particular lobby in the worker node, returns information about player role.
  rpc JoinLobby (LobbyInfo) returns (PlayerInfo) {}

  rpc GetMessages (LobbyInfo) returns (stream Message) {}

  rpc SendMessage (LobbyMessage) returns (Status) {}

  // RPC for the game admin to start the game.
  rpc StartGame (GameInfo) returns (SecretWords) {}

  rpc GetStatus (Status) returns (Status) {}
}

message LobbyMessage {
  int64 lobby_id = 1;
  string sender_id = 2;  
  string content = 3; 
}

message ChannelMessageRequest {
  string lobby_id = 1;
  string user_id = 2;
}

message Message {
  string sender_id = 1;
  string content = 2; 
  int64 timestamp = 3;

}

message Status {
  bool success = 1;
  string message = 2;
}

message LobbyInfo {
  int64 lobby_id = 1;
  string user_id = 2;
}

message PlayerInfo {
  int64 player_role = 1;
}

message SecretWords {
  string word = 1;
}

message GameInfo {
  bool start = 1;
  int64 lobby_id = 2;
}